### ISSUE TITLE
Create a better `prefer-template` rule

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#78]
As usual the built-in one is flawed and as usual the ESLint team refuses to acknowledge that. Luckily we can just build something better. Ref: https://github.com/eslint/eslint/issues/6572#issuecomment-229927387

**Comment by jfmengels** [comment]
I think it's too bad to limit this rule to only work for identifiers, but I agree that the core is too forceful at times. What do you think about looking at the length of the expression and report when using expressions (that are more complex than just a single variable) that use less than X characters (10, 15, 20?)?

This would result in something like this

### Fail

```js
// if limit set at 15, `blue('foo')` takes 11 chars
var foo = red('I') + ' like ' + blue('foo') + ' and ' + foo(a, b);
```

### Pass

```js
var foo = chalk.red('I') + ' like ' + chalk.blue('unicorns') + ' and ' + chalk.yellow('ponies');
```

I'm fine with this rule, but would love it if stuff like `'prefix' + format(a, b)` would be reported, as I'd prefer a template to be used here.

**Comment by sindresorhus** [comment]
@jfmengels Length is very arbitrary though. I don't see why the above examples should be treated differently. It's not just the length that makes some template literals hard to read. Usually it's more about the complexity of the thing inside the template braces.

We could use multiple heuristics:

- No expressions.
- Only one property deep is allowed: `foo.bar`, but not `foo.bar.baz`.
- If a function call, max 2 arguments, and if string literals, the combined max length of `x`. (What should `x` be?)

Thoughts?

> I'm fine with this rule, but would love it if stuff like 'prefix' + format(a, b) would be reported, as I'd prefer a template to be used here.

Me too, but hard to find an automatic balance here. Let's think about it some more.

**Comment by florianb** [comment]
When trying to recognize string assembling expressions i takes me more effort to understand it the more the string is distracted by complexity.

Probably it is possible to estimate the distraction by the proportion of plain text and amount of expressions. I guess thats a little bit weird.. 😄

**Comment by sindresorhus** [comment]
Let's start with this heuristic and we can improve it based on actual usage:

- No expressions.
- Only one property deep is allowed: `foo.bar`, but not `foo.bar.baz`.
- If a function call, max 2 arguments, and if string literal argument, max length of 20.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by sindresorhus** [comment]
I looked into this again, and I don’t think we should add this rule after all.

ESLint’s built-in `prefer-template` is still too broad for the original use case, but a Unicorn replacement would need subjective heuristics around expression complexity, property depth, function calls, argument length, and similar cases. That feels too fragile and too low-demand to justify maintaining our own version.
